test/hotspot/jtreg/compiler/c2/irTests/TestScopedValue.java

```diff
@@ -0,0 +1,708 @@
+        for (int i = 0; i < 10_000; i++) {
+            dst[i] = src[i] * sv.get().getValue();
+        }
+    }
+
+    @Run(test = "testFastPath14", mode = RunMode.STANDALONE)
+    private void testFastPath14Runner() throws Exception {
+        double[] array = new double[10_000];
+        ScopedValue.where(sv, new MyDouble(42)).run(
+                () -> {
+                    MyDouble unused = sv.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        testFastPath14(array, array);
+                    }
+                });
+        forceCompilation("testFastPath14", double[].class, double[].class);
+    }
+
+    // Check uncommon trap is recorded at the right byte code (a cache miss) so on re-compilation,
+    // the cache null check still branches to an uncommon trap
+    @Test
+    @IR(counts = {IRNode.UNSTABLE_IF_TRAP, ">= 1" })
+    public static Object testFastPath15() {
+        return svObject.get();
+    }
+
+    @Run(test = "testFastPath15", mode = RunMode.STANDALONE)
+    private void testFastPath15Runner() throws Exception {
+        // Profile data will report a single of the 2 cache locations as a hit
+        runAndCompile15();
+        // Force a cache miss
+        ScopedValue.where(svObject, new Object()).run(
+                () -> {
+                    testFastPath15();
+                });
+        Method m = TestScopedValue.class.getDeclaredMethod("testFastPath15");
+        TestFramework.assertDeoptimizedByC2(m);
+        // Compile again
+        runAndCompile15();
+    }
+
+    private static void runAndCompile15() throws NoSuchMethodException {
+        ScopedValue.where(svObject, new Object()).run(
+                () -> {
+                    Object unused = svObject.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        testFastPath15();
+                    }
+                });
+        forceCompilation("testFastPath15");
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_D, "1", IRNode.CALL_OF_METHOD, "slowGet", "1" })
+    public static double testSlowPath1() {
+        ScopedValue<MyDouble> localSV = sv;
+        MyDouble sv1 = localSV.get();
+        MyDouble sv2 = localSV.get(); // should optimize out
+        return sv1.getValue() + sv2.getValue();
+    }
+
+    @Run(test = "testSlowPath1")
+    private void testSlowPath1Runner() throws Exception {
+        ScopedValue.where(sv, new MyDouble(42)).run(
+                () -> {
+                    if (testSlowPath1() != 42 + 42) {
+                        throw new RuntimeException();
+                    }
+                });
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_D, "1", IRNode.CALL_OF_METHOD, "slowGet", "1" })
+    public static double testSlowPath2() {
+        ScopedValue<MyDouble> localSV = sv;
+        MyDouble sv1 = localSV.get();
+        notInlined();
+        MyDouble sv2 = localSV.get(); // should optimize out
```

Comment:
Why does this work now, and some other cases with `notInlined` in between do not work?

Reply:
Because the `sv` field is loaded in a local before the call.